activesupport/lib/active_support/file_evented_update_checker.rb

```diff
@@ -0,0 +1,71 @@
+require 'listen'
+  module ActiveSupport
+    class FileEventedUpdateChecker
+      attr_reader :listener
+      def initialize(files, dirs={}, &block)
+      	@files = Set.new
+      	@files = files.map { |f| File.expand_path(f)}
+      	@dirs = Hash.new
+      	dirs.map do |key,value|
+          @dirs[File.expand_path(key)] = Array(value) if !Array(value).empty?
+      	end
+      	@block = block
+      	@modified = false
+      	@listener = Listen.to(*base_directories, relative_paths: false)
+      	@listener.change(&method(:changed))
+      	@listener.start
+      end
+
+      def updated?
+        @modified
+      end
+
+      def execute
+        @block.call
+      ensure
+        @modified = false
+      end
+
+      # Execute the block given if updated.
+      def execute_if_updated
+        if updated?
+          execute
+          true
+        else
+          false
+        end
+      end
+
+      def watching?(file)
+        return true if @files.include?(file)
+        cfile = file
+        while !cfile.eql? "/"
```

Comment:
Here the main aim is to check that file is among the watched one or not. So first I am checking that it exists in @files, and then for checking in directory instead of iterating through all the entries present in @dirs I am checking prefix of file name is there in @dirs and then extension is among the associated ones for that directory.